main.py

```diff
@@ -574,6 +580,14 @@ def notification_handler(self, response):
         elif method == "window/logMessage" and log_server:
             server_log(self.process.args[0],
                        response.get("params").get("message"))
+        elif method == "rustDocument/diagnosticsBegin":
```

Comment:
What are all these `rustDocument/*` methods? Where are they documented?

Reply:
these are LSP extensions specific to the Rust Language Server implementation. They're [documented here.](https://github.com/rust-lang-nursery/rls/blob/master/contributing.md#extensions-to-the-language-server-protocol)

Reply:
They are custom messages specific to RLS which any LSP consumer is free to ignore or handle.

These Rust ones are probably the most commonly used ones - the initial 'analyse' build can take a long time, the VS extension displays a spinner with some text to assure the user that things are happening :)

This specific handling should be regarded as a future feature for the LSP API - how can we expose hooks for an "LSP-contrib-rust" package to add extra handlers and features?

Reply:
Cool! I think this instigates an interesting discussion on how LSP should handle server-specific code.